server/src/updaters/relationship-updaters.js

```diff
@@ -0,0 +1,106 @@
+import { fetchFriendRequestRelationshipOperations } from '../fetchers/relationship-fetchers';
+import { dbQuery, SQL } from '../database';
+
+async function updateRelationships(
+  viewer: Viewer,
+  request: RelationshipRequest,
+) {
+  const { action } = request;
+
+  if (!viewer.loggedIn) {
+    throw new ServerError('not_logged_in');
+  }
+
+  const uniqueUserIDs = [...new Set(request.userIDs)];
+  const users = await fetchUserInfos(uniqueUserIDs);
+
+  const invalid_user: string[] = [];
+  const userIDs: string[] = [];
+  for (let userID of uniqueUserIDs) {
+    if (userID === viewer.userID || !users[userID].username) {
+      invalid_user.push(userID);
+    } else {
+      userIDs.push(userID);
+    }
+  }
+
+  if (action === 'send_friend_request') {
+    const {
+      userRelationshipOperations,
+      errors: friendRequestErrors,
+    } = await fetchFriendRequestRelationshipOperations(viewer, userIDs);
+
+    const undirectedInsertRows = [];
+    const directedInsertRows = [];
+    const directedDeleteIDs = [];
+    for (const userID in userRelationshipOperations) {
+      const operations = userRelationshipOperations[userID];
+      const ids = [userID, viewer.userID].map(Number).sort((a, b) => a - b);
+
+      for (const operation of operations) {
+        if (operation === 'delete_directed') {
+          directedDeleteIDs.push(userID);
+        } else if (operation === 'friend') {
+          undirectedInsertRows.push([...ids, undirectedStatus.FRIEND]);
+        } else if (operation === 'pending_friend') {
+          const status = directedStatus.PENDING_FRIEND;
+          directedInsertRows.push([viewer.userID, userID, status]);
+        } else if (operation === 'know_of') {
+          undirectedInsertRows.push([...ids, undirectedStatus.KNOW_OF]);
+        } else {
+          invariant(false, `unexpected relationship operation ${operation}`);
+        }
+      }
+    }
+
+    const undirectedInsertQuery = SQL`
+      INSERT INTO relationships_undirected (user1, user2, status)
+      VALUES ${undirectedInsertRows}
+      ON DUPLICATE KEY UPDATE status = MAX(status, VALUES(status))
+    `;
+    const directedInsertQuery = SQL`
+      INSERT INTO relationships_directed (user1, user2, status)
+      VALUES ${directedInsertRows}
+      ON DUPLICATE KEY UPDATE status = VALUES(status)
+    `;
+    const directedDeleteQuery = SQL`
+      DELETE FROM relationships_directed
+      WHERE 
+        (user1 = ${viewer.userID} AND user2 IN (${directedDeleteIDs})) OR
+        (status = ${directedStatus.PENDING_FRIEND} AND 
+          user1 IN (${directedDeleteIDs}) AND user2 = ${viewer.userID})
+    `;
+
+    const promises = [
+      dbQuery(undirectedInsertQuery),
+      dbQuery(directedInsertQuery),
+    ];
+
+    if (directedDeleteQuery.length) {
+      promises.push(dbQuery(directedDeleteQuery));
+    }
+
+    await Promise.all(promises);
+
+    return { ...friendRequestErrors, invalid_user };
```

Comment:
Nit: I like that in your old approach, you avoid settings any error keys unless there actually are errors. In this case, we'll always be returning an empty `invalid_users` array, which isn't a big deal, but would probably be cleaner/more consistent if we didn't

Reply:
I agree, I've changed it because I had an `inexact spread` Flow warning. To avoid that warning I'm using `Object.assign()` now